File: 1_Loading_Data&Pckgs.r
library(ncdf4)
library(raster)
library(randomForest)
library(caret)
library(rgdal)
library(sf)
library(RANN)
library(rasterVis)
library(akima)
library(dplyr)
library(grid)
library(data.table)

#MONTHLY GRACE#


setwd("D:/Research/MASTER DATA/GRACE NC/2012")

#list to hold clipped GRACE files
grace_clipped_list <- list()
grace_files <- list.files(path = ".", pattern = "GRD.*\\.nc$", full.names = TRUE)

#Loop through GRACE files and process each
for (i in 1:length(grace_files)) {
  #open
  grace_nc <- nc_open(grace_files[i])
  grace_var <- "lwe_thickness"
  grace_data <- ncvar_get(grace_nc, grace_var)
  grace_mtrx <- matrix(grace_data, ncol = nrow(grace_nc$dim$lon$vals))
  grace_rst <- raster(grace_files[i])
  
  #shift extent of GRACE raster by 180 degrees (necessary for proper alignment)
  grace_rst <- rotate(grace_rst, angle = 180, transform = TRUE)
  extent(grace_rst)
  
  #MS Embayment Shapefile
  shapefile <- st_read("D:/Research/MS_Embayment_SHP/Mississippi_embayment.shp")
  shapefile_proj <- st_transform(shapefile, st_crs(grace_rst))
  
  #cropping GRACE by shapefile
  grace_clip <- crop(grace_rst, shapefile_proj)
  
  #project into UTM
  grace_utm <- projectRaster(grace_clip, crs = "+proj=utm +zone=15 +datum=NAD83 +units=m +no_defs")
  
  #close file
  nc_close(grace_nc)
  
  #append to list
  grace_clipped_list <- append(grace_clipped_list, grace_clip)
}


#MONTHLY GLDAS 
folder_path <- "D:/Research/MASTER DATA/GLDAS/1x1/2012"
file_list <- list.files(path= folder_path, pattern = "*.nc4")

#define the variables to extract
var_names <- c("Evap_tavg", "Qs_acc", "SWE_inst", "CanopInt_inst", "RootMoist_inst", "SoilMoi0_10cm_inst", "SoilMoi10_40cm_inst", "SoilMoi40_100cm_inst", "SoilMoi100_200cm_inst", "Rainf_f_tavg")

#initialize  empty list to hold the stacked rasters for all gldas 1x1 variables
gldas1_stacked <- list()

#loop over each yearly .nc4 file
for (file_name in list.files(folder_path, pattern = "\\.nc4$", full.names = TRUE)) {
  # Initialize an empty list to hold the rasters for the current file
  var_rasters <- list()
  
  #loop over variable names and extract each one as a raster
  for (var_name in var_names) {
    # Get the raster for the current variable
    cur_raster <- raster(file_name, varname = var_name)
    # Add the raster to the list of rasters for the current file
    var_rasters[[var_name]] <- cur_raster
  }
  
  #loop over the list of rasters and crop each one
  cur_gldas1_clipped <- list()
  for (var_name in var_names) {
    cur_raster <- var_rasters[[var_name]]
    cur_raster_clipped <- crop(cur_raster, extent(shapefile_proj))
    cur_gldas1_clipped[[var_name]] <- cur_raster_clipped
  }
  
  #stack the clipped rasters for the current file
  cur_gldas1_stacked <- stack(cur_gldas1_clipped)
  
  #append the stacked rasters for the current file to the list of stacked rasters for all files
  gldas1_stacked <- append(gldas1_stacked, cur_gldas1_stacked)
}


#MONTHLY GLDAS 0.25x0.25

folder_path025 <- "D:/Research/MASTER DATA/GLDAS/0.25x0.25/2012"
file025_list <- list.files(path= folder_path, pattern = "*.nc4")
var_names <- c("Evap_tavg", "Qs_acc", "SWE_inst", "CanopInt_inst", "RootMoist_inst", "SoilMoi0_10cm_inst", "SoilMoi10_40cm_inst", "SoilMoi40_100cm_inst", "SoilMoi100_200cm_inst", "Rainf_f_tavg")

gldas025_stacked <- list()

#loop over each yearly .nc4 file
for (file_name in list.files(folder_path025, pattern = "\\.nc4$", full.names = TRUE)) {
  # Initialize an empty list to hold the rasters for the current file
  var_rasters <- list()
  
  #loop over variable names and extract each one as a raster
  for (var_name in var_names) {
    # Get the raster for the current variable
    cur_raster <- raster(file_name, varname = var_name)
    # Add the raster to the list of rasters for the current file
    var_rasters[[var_name]] <- cur_raster
  }
  #loop over the list of rasters and crop each one
  cur_gldas025_clipped <- list()
  for (var_name in var_names) {
    cur_raster <- var_rasters[[var_name]]
    cur_raster_clipped <- crop(cur_raster, extent(shapefile_proj))
    cur_gldas025_clipped[[var_name]] <- cur_raster_clipped
  }
  
  cur_gldas025_stacked <- stack(cur_gldas025_clipped)
  
  #append the stacked rasters for the current file to the list of stacked rasters for all files
  gldas025_stacked <- append(gldas025_stacked, cur_gldas025_stacked)
}

File: python/1_Loading_Data&Pckgs.py
import os
import numpy as np
import rasterio
from rasterio.warp import calculate_default_transform, reproject, Resampling
from rasterio.mask import mask
from rasterio.transform import from_origin
import geopandas as gpd
from netCDF4 import Dataset

# Set working directory
os.chdir("/Users/brian/Projects/GRACE-refined/GRACE_Comp/")

# List to hold clipped GRACE files
grace_clipped_list = []
grace_files = [f for f in os.listdir(".") if f.startswith("GRD") and f.endswith(".nc")]

# Loop through GRACE files and process each
for grace_file in grace_files:
    # Open
    grace_nc = Dataset(grace_file)
    grace_var = "lwe_thickness"
    grace_data = grace_nc.variables[grace_var][:]
    grace_mtrx = np.matrix(grace_data)
    src = rasterio.open(grace_file)
    # Read raster
    """ with rasterio.open(grace_file) as src:
        #transform = from_origin(lon_bounds[0], lat_bounds[1], ds.lon.diff('lon').mean().item(), ds.lat.diff('lat').mean().item())
        if src.count == 0:
            print(f"No bands found in {grace_file}")
            continue
        grace_rst = src.read(1)
        transform = src.transform
    
    # Shift extent of GRACE raster by 180 degrees (necessary for proper alignment)
    """
    crs = 4326 # WGS 84
    transform, width, height = calculate_default_transform(
        crs, crs, src.width, src.height, *src.bounds, dst_width=src.width, dst_height=src.height
    )
    kwargs = src.meta.copy()
    kwargs.update({
        'crs': src.crs,
        'transform': transform,
        'width': width,
        'height': height
    })
    
    """ with rasterio.open('rotated.tif', 'w', **kwargs) as dst:
        reproject(
            source=rasterio.band(src, 1),
            destination=rasterio.band(dst, 1),
            src_transform=src.transform,
            src_crs=src.crs,
            dst_transform=transform,
            dst_crs=src.crs,
            resampling=Resampling.nearest
        ) """
    
    # MS Embayment Shapefile
    shapefile = gpd.read_file("/Users/brian/Projects/GRACE-refined/MS_Embayment_SHP/Mississippi_embayment.shp")
    shapefile_proj = shapefile.to_crs(crs)
    
    # Cropping GRACE by shapefile
    """ with rasterio.open('rotated.tif') as src:
        out_image, out_transform = mask(src, shapefile_proj.geometry, crop=True)
        out_meta = src.meta.copy()
        out_meta.update({"driver": "GTiff",
                         "height": out_image.shape[1],
                         "width": out_image.shape[2],
                         "transform": out_transform}) """
    out_image, out_transform = mask(src, shapefile_proj.geometry, crop=True)
    out_meta = src.meta.copy()
    with rasterio.open("clipped.tif", "w", **out_meta) as dest:
        dest.write(out_image, 1)
    
    # Append to list
    grace_clipped_list.append(out_image)
    grace_nc.close()

# MONTHLY GLDAS 1 degree by 1 degree resolution
folder_path = "/Users/brian/Projects/GRACE-refined/GLDAS_1x1/"
var_names = ["Evap_tavg", "Qs_acc", "SWE_inst", "CanopInt_inst", "RootMoist_inst", "SoilMoi0_10cm_inst", "SoilMoi10_40cm_inst", "SoilMoi40_100cm_inst", "SoilMoi100_200cm_inst", "Rainf_f_tavg"]

gldas1_stacked = []

# Loop over each yearly .nc4 file
for file_name in [f for f in os.listdir(folder_path) if f.endswith(".nc4")]:
    var_rasters = {}
    
    # Loop over variable names and extract each one as a raster
    with Dataset(os.path.join(folder_path, file_name)) as ds:
        for var_name in var_names:
            var_rasters[var_name] = ds.variables[var_name][:]
    
    cur_gldas1_clipped = {}
    for var_name in var_names:
        cur_raster = var_rasters[var_name]
        with rasterio.open(os.path.join(folder_path, file_name)) as src:
            out_image, out_transform = mask(src, shapefile_proj.geometry, crop=True)
            cur_gldas1_clipped[var_name] = out_image
    
    gldas1_stacked.append(cur_gldas1_clipped)

# MONTHLY GLDAS at 0.25x0.25 resolution
folder_path025 = "/Users/brian/Projects/GRACE-refined/GLDAS_025x025/"
gldas025_stacked = []

# Loop over each yearly .nc4 file
for file_name in [f for f in os.listdir(folder_path025) if f.endswith(".nc4")]:
    var_rasters = {}
    
    # Loop over variable names and extract each one as a raster
    with Dataset(os.path.join(folder_path025, file_name)) as ds:
        for var_name in var_names:
            var_rasters[var_name] = ds.variables[var_name][:]
    
    cur_gldas025_clipped = {}
    for var_name in var_names:
        cur_raster = var_rasters[var_name]
        with rasterio.open(os.path.join(folder_path025, file_name)) as src:
            out_image, out_transform = mask(src, shapefile_proj.geometry, crop=True)
            cur_gldas025_clipped[var_name] = out_image
    
    gldas025_stacked.append(cur_gldas025_clipped)

print(gldas025_stacked)